### PR TITLE
MU07 - Chainlink Relayers

### PR DESCRIPTION
**Pull request by chapati23**
## MU07 Overview

1. Deploy ChainlinkRelayerFactory proxied with a ProxyAdmin owned by Mento Labs multisig.
2. Deploy ChainlinkRelayer instances for two rate feeds: `relayed:CELOPHP` and `relayed:PHPUSD`.
3. Submit a Celo Governance Proposal that:
   - Whitelists the relayers for the two rate feeds.
   - Sets the `relayed:CELOPHP` as the "equivalent token" for the PSO token address. Meaning that when paying with gas, the blockchain node will use the `relayed:CELOPHP` rate feed when gas is payed in the PSO token.

### Testing / development helpers

In order to have as close a structure as possible on testnets, this PR also deploys a few `MockChainlinkAggregator` contracts:
- `PHPUSD` to Alfjores
- `PHPUSD` to Baklava
- `CELOUSD` to Baklava

Using these on Alfajores we have:
- `relayed:CELOPHP` built from chainlink's own testnet aggregator of `CELOUSD` and our mock aggregator of `PHPUSD`
- `relayed:PHPUSD` built from our mock aggregator of `PHPUSD`

And on Baklava we have everything based only on our mock aggregators.

#### RelayerStatus

```bash
$ yarn script:dev -n alfajores -s RelayerStatus
```

Get a status output for all defined relayers:
```
  ====== CELO/PHP (CELO/USD:USD/PHP) =======
  RateFeedID: 0xaFc02368A174Cd08e01c373de6D0B537CECF43C8
  Num rates: 1
  Median rate: 25850491393557810218000000
  ====== PHP/USD =======
  RateFeedID: 0xab921d6ab1057601A9ae19879b111fC381a2a8E9
  Num rates: 1
  Median rate: 17540930000000000000000
```

#### RelayerReport

```bash
$ yarn script:dev -n alfajores -s RelayerReport
```

Trigger report on all relayers.

```
Relayed CELO/PHP (CELO/USD:USD/PHP) successfully.
Relayed PHP/USD successfully.
```

#### UpdateMockChainlinkAggregators

```
$ yarn script:dev -n alfajores -s UpdateMockChainlinkAggregators
```

Update the mock aggregators on testnets with values from the mainnet aggregators that they mirror.

## Other Changes

- Removed the `forge_script` helper functions because it got out of hand and it was only used in a couple of places
- Improved the `verify:celoscan` command to check ABIs on celoscan first, and skip verification.

## Related Issues

- https://github.com/mento-protocol/mento-general/issues/459

### CODE REVIEW

File: bin/deploy.sh
Comment (chapati23) [inline bin/deploy.sh:37]: a fellow emoji padder, i appreciate it

File: bin/deploy.sh
Comment (chapati23) [inline bin/deploy.sh:1]: haven't worked with these scripts yet, just a general shell hygiene question: why no `set -euo pipefail`?

just in case:

1. `set -e`: Exit immediately if a command exits with a non-zero status.
2. `set -u`: Treat unset variables as an error when substituting.
3. `set -o pipefail`: The return value of a pipeline is the status of the last command to exit with a non-zero status, or zero if no command exited with a non-zero status.

File: script/deploy/DeployChainlinkRelayerFactory.sol
Comment (chapati23) [inline script/deploy/DeployChainlinkRelayerFactory.sol:45]:
```suggestion
    console.log("ChainlinkRelayerFactoryProxyAdmin: ", address(proxyAdmin));
```

File: script/deploy/DeployChainlinkRelayers.sol
Comment (chapati23) [inline script/deploy/DeployChainlinkRelayers.sol:45]: grouping-wise would feel more logical to move PHP/USD above CELO/PHP, then it'd read:
1. Deploy CELO/USD
2. Deploy PHP/USD
3. Compose both feeds into CELO/PHP

Comment (chapati23) [inline script/deploy/DeployChainlinkRelayers.sol:45]: Why does this `rateFeed` not have the `chainlink:` prefix?

File: script/deploy/DeployChainlinkRelayers.sol
Comment (chapati23) [inline script/deploy/DeployChainlinkRelayers.sol:71]: tbh i didn't fully understand the `IfNoneOrDifferent` suffix here, could you elaborate?

Comment (chapati23) [reply script/deploy/DeployChainlinkRelayers.sol:71]: ok, think i got it now, only deploy if no current relayer exists in factory storage OR if a relayer exists but with a different address. off the top of my head don't have a better naming suggestion but feels like there could be one :))

File: script/deploy/WhitelistRelayers.sol
Comment (chapati23) [inline script/deploy/WhitelistRelayers.sol:66]: can't connect the comment to the `createProposal` line below, might be missing something 🤔

File: script/upgrades/dependencies.json
Comment (chapati23) [inline script/upgrades/dependencies.json:24]: are we missing `Chainlink.CELOPHP` here or left out on purpose?

File: bin/deploy.sh
Comment (bowd) [reply bin/deploy.sh:1]: https://github.com/mento-protocol/mento-deployment/blob/9a4fecf4f8910bf8a5171b515b7e7237c5a477bb/bin/setup.sh#L1

We do it, but in the shared setup file, because our bash scripts have functions and are cool. 😂

File: script/upgrades/dependencies.json
Comment (bowd) [reply script/upgrades/dependencies.json:24]: CELOPHP doesn't exist on Chainlink, these are the addresses of Chainlink aggregators, and we're combining CELOUSD and PHPUSD to report CELOPHP

File: bin/deploy.sh
Comment (chapati23) [reply bin/deploy.sh:1]: interesting, wasn't aware that these flags apply even to the calling script. was always naively assuming they're locally scoped per file. TIL something about bash :)

Comment (bowd) [reply bin/deploy.sh:1]: It's true that they're scoped per script, but in this case we're sourcing the file, so we're basically executing all the commands there in the context of each script.

**Comment by bowd** [comment]
I tried running the proposal on Baklava but it seams that the cLabs contracts team have also abandoned Baklava and SortedOracles is not up to date with the equivalent token implementation :(

File: script/dev/dev-RelayerStatus.sol
Comment (chapati23) [inline script/dev/dev-RelayerStatus.sol:26]:
```suggestion
  private ISortedOracles sortedOracles;
```

File: script/dev/dev-RelayerStatus.sol
Comment (chapati23) [inline script/dev/dev-RelayerStatus.sol:27]:
```suggestion
  private ChainlinkRelayerFactory relayerFactory;
```

File: script/dev/dev-DeployMockChainlinkAggregator.sol
Comment (philbow61) [inline script/dev/dev-DeployMockChainlinkAggregator.sol:17]: is the PHPUSD supposed to be in the contract name?

File: script/upgrades/MU07/MU07Checks.sol
Comment (philbow61) [inline script/upgrades/MU07/MU07Checks.sol:1]: nit: we could add some checks here verifying, that the ChainlinkRelayerFactory, Relayers setup/deployment worked as expected: 
- proxy implementation set correctly 
- proxy admin set correctly 
- only 2 Relayers

File: script/upgrades/MU07/MU07Checks.sol
Comment (philbow61) [inline script/upgrades/MU07/MU07Checks.sol:51]: could we combine the if with the assertEq in one require?

File: script/upgrades/MU07/MU07Checks.sol
Comment (philbow61) [inline script/upgrades/MU07/MU07Checks.sol:28]: I guess this will be StableTokenPHPProxy again

File: script/dev/dev-RelayerReport.sol
Comment (bowd) [reply script/dev/dev-RelayerReport.sol:22]: Yeah we have the warnings locally as well, we just ignored them :(

File: script/upgrades/MU07/MU07Checks.sol
Comment (bowd) [reply script/upgrades/MU07/MU07Checks.sol:1]: I felt like it would be overkill, I think we should limit the Checks to verify the effects of the governance proposal, because those are a bit more "uncertain", because what we do needs to pass through a different system. If we're checking the relayer setup we're basically checking the results of a straightforward script so I'm not sure I would add that imo.

File: script/upgrades/MU07/MU07Checks.sol
Comment (bowd) [reply script/upgrades/MU07/MU07Checks.sol:53]: we could but I've started to dislike the `abi.encodePacked` string concatenation thing in order to build a useful revert reason here for printing, so I decided to use the logs instead.

File: broadcast/ExecuteProposal.sol/44787/MU07.json
Comment (bowd) [reply broadcast/ExecuteProposal.sol/44787/MU07.json:1]: ExecuteProposal is the foundry script that executes CGPs. I think it's worth keeping them in the repo as standard so that we have those references. We kept it for MUGOV because it also deployed contracts, but maybe future ones will also deploy contracts, so it would be good to record these.

File: script/upgrades/MU07/MU07.sol
Comment (chapati23) [inline script/upgrades/MU07/MU07.sol:163]:
```suggestion
   * canonical id: `relayed:CELOPHP`, and then have address(PUSO) point to that for
```

**Review by chapati23 — COMMENTED** [review]
one last typo, then good to go from my side

**Review by bowd — APPROVED** [review]

File: script/upgrades/MU07/deploy/MU07-Deploy-ChainlinkRelayerFactory.sol
Comment (nvtaveras) [inline script/upgrades/MU07/deploy/MU07-Deploy-ChainlinkRelayerFactory.sol:23]: address looks correct, but maybe get it from `contracts.dependency("MentoLabsMultisig")` ?

File: script/upgrades/dependencies.json
Comment (nvtaveras) [inline script/upgrades/dependencies.json:69]: I think the latest one here is `0x995c827326Bc7393BFA28785f98744548874B2f0`? 

That's what I got from running the `UpdateMockChainlinkAggregators` script

File: script/utils/Script.sol
Comment (nvtaveras) [inline script/utils/Script.sol:40]: nit: dupped from `script/utils/mento/Oracles.sol`

File: script/upgrades/dependencies.json
Comment (bowd) [reply script/upgrades/dependencies.json:69]: True, I didn't update it in my last deployment, grr.

File: script/utils/Script.sol
Comment (bowd) [reply script/utils/Script.sol:40]: duped because importing floating functions doesn't work in solidity v0.5 so they're not compatible. It will be nicer after I finish my OTHER BIG ONE to refactor the whole deployments repo 😂

**Comment by chapati23** [comment]
Merging currently blocked until this is merged https://github.com/celo-org/governance/pull/472

**Review by nvtaveras — APPROVED** [review]

**Review by philbow61 — APPROVED** [review]

**Pull request merged**
